refactor: remove ABI decoding from pxe private events

mverzilli opened the pull request:
Part of F-96

This refactor moves the responsibility of ABI-decoding private events from PXE to wallet. There are some details/design decisions to iron out though, I'll try to summarize them below.

## `PrivateEvent`: to `InBlock` or not to `InBlock`

Before this change, `PXE#getPrivateEvents` was generic on the type of event, so we didn't need an explicit return type. After this change, PXE returns `PrivateEvent`'s that the wallet must ABI decode. 

There's an open conversation regarding whether `PrivateEvent` should be defined in terms of `InBlock`. I'm a bit on the fence about forcing an artificial `data` attribute on the structure, so I'm discussing potentially changing `InBlock<T>` from a generic to a plain type that can be used in intersections (hence more degrees of freedom for whoever is defining types on top of it). If that approach is blessed I'll probably tackle it in its own PR since `InBlock` is defined in stdlib and changing it could ripple. 

## Unit test for PXE#getPrivateEvents

I wanted to have some tests that don't require an e2e harness to check basic functioning. Most of the setup of the test is mocking context so that calling  `sync_private_state` doesn't blow up.

The rest of it leverages `PrivateDataEventDataProvider` to make sure some events are ready to be returned.

This at least works so if we break interfaces there's a quick unit test failing to show it. More granular behaviors can be better tested at the `PrivateDataEventDataProvider` unit suite, and less artificial examples can be exercised on e2e.

## Unit test for ABI decoding at BaseWallet#getPrivateEvents

Similarly, I wanted to have some unit test in place before proceeding to refactor these APIs further. At this layer I just wanted to have some checks that we're actually ABI-decoding. I couldn't find a straightforward/generic way to encode event data so I'm manually serializing it. If we prefer so, I can write an ad-hoc contract for this test but first I would like to validate the general approach.

nventuro commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -73,14 +74,19 @@ import { AddressDataProvider } from './storage/address_data_provider/address_dat
 import { Synchronizer } from './synchronizer/index.js';
 
+export type PrivateEvent = {
+  msgContent: Fr[];
```
```suggestion
  packedEvent: Fr[];
```

Terminology nit. 'messages' are the things people send to one another (through various channels), they have a ciphertext (and corresponding plaintex) and some metadata (tx hash etc). Some messages are event messages, which are identified by the 'msg id' field in the message header inside the message plaintext. An event message contains the packed content (an array of fields), which can be decoded to the actual typed event data, and the selector.

Should this struct not have the selector btw?

mverzilli replied:
ah yes, it's a leaky abstraction indeed, will fix. and re: selector, yes, probably a good idea. I'm introducing an EventFilter next and I guess there will end up being some symmetry between that type and this one in its "final" form

benesjan commented on yarn-project/pxe/src/pxe.ts:
```suggestion
   * @param eventSelector - Identifies the event type we are requesting.
```

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -1072,13 +1078,13 @@ export class PXE {
    * @param recipients - The addresses that decrypted the logs.
    * @returns - The deserialized events.
```
```suggestion
   * @returns - The events.
```

We now don't perform decoding here so that was a lie.

i think it would be fine to altogether drop the return value description here as it's clear.

benesjan reviewed (approved):
LGTM

nventuro reviewed (approved):
Lovely!

mverzilli committed:
refactor: remove ABI decoding from pxe private events

Part of F-96

This refactor moves the responsibility of ABI-decoding private events from PXE to wallet. There are some details/design decisions to iron out though, I'll try to summarize them below.

## `PrivateEvent`: to `InBlock` or not to `InBlock`

Before this change, `PXE#getPrivateEvents` was generic on the type of event, so we didn't need an explicit return type. After this change, PXE returns `PrivateEvent`'s that the wallet must ABI decode.

There's an open conversation regarding whether `PrivateEvent` should be defined in terms of `InBlock`. I'm a bit on the fence about forcing an artificial `data` attribute on the structure, so I'm discussing potentially changing `InBlock<T>` from a generic to a plain type that can be used in intersections (hence more degrees of freedom for whoever is defining types on top of it). If that approach is blessed I'll probably tackle it in its own PR since `InBlock` is defined in stdlib and changing it could ripple.

## Unit test for PXE#getPrivateEvents

I wanted to have some tests that don't require an e2e harness to check basic functioning. Most of the setup of the test is mocking context so that calling  `sync_private_state` doesn't blow up.

The rest of it leverages `PrivateDataEventDataProvider` to make sure some events are ready to be returned.

This at least works so if we break interfaces there's a quick unit test failing to show it. More granular behaviors can be better tested at the `PrivateDataEventDataProvider` unit suite, and less artificial examples can be exercised on e2e.

## Unit test for ABI decoding at BaseWallet#getPrivateEvents

Similarly, I wanted to have some unit test in place before proceeding to refactor these APIs further. At this layer I just wanted to have some checks that we're actually ABI-decoding. I couldn't find a straightforward/generic way to encode event data so I'm manually serializing it. If we prefer so, I can write an ad-hoc contract for this test but first I would like to validate the general approach.
